### ISSUE TITLE
Linter rule ideas

### ISSUE THREAD

**Issue by scripthunter7** [AdguardTeam/AGLint#13]
In this issue, we can collect the main linter rule ideas. A linter rule basically targets a specific problem and the rules can be enabled / disabled individually in the linter config. Let's try to define each rule as best as possible.

Example rules:
- Check unopened / unclosed if directives. For example:
  ```adblock
  ! Unopened if:
  !#endif
  rule
  ! Unclosed if:
  !#if (adguard)
  rule
  ```
- Warning on multilple selectors. For example:
  ```adblock
  ! Multiple selectors are used simultaneously within the rule:
  example.com##.ad, .ad2
  ! But this is ok:
  example.net###.ad
  ```
  - Suggest fix (the linter can suggest a fix):
    ```adblock
    example.com##.ad
    example.com##.ad2
    example.net###.ad
    ```
- Check redundant wildcards. For example: 
  ```adblock
  ! Redundant *^
  example.*^
  ! Redundant *
  ex**ample.com
  ```

@ameshkov 
@Alex-302

**Comment by jellizaveta** [comment]
Check on the number of vertical lines in rules to avoid such typos:

<details><summary>Screenshot:</summary>

<img width="526" alt="Screenshot 2023-05-25 at 14 29 32" src="https://github.com/AdguardTeam/AGLint/assets/91964807/c9474ab9-1500-4076-ba0d-2eec8949cbc7">

<img width="587" alt="Screenshot 2023-05-25 at 14 29 55" src="https://github.com/AdguardTeam/AGLint/assets/91964807/8476314a-641a-41da-a922-da269b38b48a">

</details><br/>

**Comment by piquark6046** [comment]
Check and return an error or warn if a filter maintainer writes a regexp filtering rule that is not supported in an adblocker.

Bad:
```adblock
!#if ( adguard_app_ios || adguard_ext_safari )
/^https?:\/\/example\.com\/[A-z0-9]{10,}/
!#endif
```

Good:
```adblock
!#if ( !adguard_app_ios && !adguard_ext_safari )
/^https?:\/\/example\.com\/[A-z0-9]{10,}/
!#endif
```

https://webkit.org/blog/3476/content-blockers-first-look/
https://developer.apple.com/documentation/safariservices/creating_a_content_blocker

**Comment by Alex-302** [comment]
@piquark6046 

> ```
> {10,}
> ```

is not supported in Safari CB.

**Comment by piquark6046** [comment]
Edited.

**Comment by Alex-302** [comment]
@scripthunter7 it is better to generate an error, because in this case the rule will not work (any - both scriptlets and hide/CSS rules)

![image](https://github.com/AdguardTeam/AGLint/assets/8361299/a2f34d20-5335-4376-a2c9-8e9b9975426d)

The question arose because of the addition of a rule with these quotes to the filter (has been corrected by the author).

**Comment by piquark6046** [comment]
Add a ReDoS protection feature that checks a regexp filter.
Reference: https://github.com/AdguardTeam/FiltersCompiler/issues/162

**Comment by piquark6046** [comment]
Support [action operator of uBlock Origin](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#action-operators).
For example, each of the following syntax are same:
Meaning | AdGuard | uBlock Origin
------- | ------- | ------------
Modifying style of an element | `domain#$#element { style }` | `domain##element:style( style )`
Removing class property of an element | `domain#%#//scriptlet('remove-class', ...)` | `domain##element:remove-class(...)`

**Comment by scripthunter7** [comment]
Detection of redundant rules. It can be a rather complex and resource-intensive topic. For example, if the list has the following rule:
```adblock
##.ad
```
then the following rule becomes unnecessary:
```adblock
example.net##.article + section.ad
```
since the previous rule is "stronger" / "more general".

Following a similar analogy, this is also true for network rules:
```adblock
! this rule
/ads.js^$script
! makes this rule redundant:
||example.org/ads.js^$third-party,script
```

**Comment by Alex-302** [comment]
Add protection against accidental insertion of words, similar to generic rules without modifiers, special characters, dots etc. Only letters and digits.
For example recent case https://github.com/AdguardTeam/AdguardFilters/blob/4c4128acb68547e7c3c7a3f79c49e454457dd077/BaseFilter/sections/foreign.txt#L2159
It may not be necessary to limit the length. Such rules are not used in filters.

**Comment by Alex-302** [comment]
*for https://github.com/AdguardTeam/VscodeAdblockSyntax/issues/98*

Detect space in domains list

<details>
<summary>Details</summary>

![image](https://github.com/AdguardTeam/AGLint/assets/8361299/25ac4891-4cd5-4c41-b1b3-8eda1e58a839)

</details>

```
/firebase.js$domain=example.com | example.org
/firebase.js$domain=example.com| example.org
/firebase.js$domain=example.com |example.org
example.com , example.org##body
example.com, example.org##body
example.com ,example.org##body
```

**Comment by jellizaveta** [comment]
Add validation for the popup modifier. Only one domain can be added.   
<details><summary>Screenshot:</summary>

<img width="835" alt="Screenshot 2023-09-12 at 17 52 20" src="https://github.com/AdguardTeam/AGLint/assets/91964807/cc08cdc0-fbff-4f69-97f2-db00067bd44a">

</details><br/>

**Comment by piquark6046** [comment]
The configurable linter rules that allow/reject type of a filter by a file-level or scope-level.


```adblock
! aglin "reject-type" ["extcss"]
example.com##.ad
example.com##.ad2
! aglin
```
will be passed. But, the following filter will be rejected:
```adblock
! aglin "reject-type" ["extcss"]
example.com#?#.ad
example.com##.ad2
! aglint
```
If you want to allow/reject type of a filter by a file-level, you can use the following rule:
```adblock
! aglin "reject-type-file" ["extcss"]
```

**Comment by Alex-302** [comment]
`$permissions` rules with argument with `()` detected as invalid

Example:
`||browserleaks.com^$permissions=geolocation()`

![image](https://github.com/AdguardTeam/AGLint/assets/8361299/37a3989b-e601-4ae9-85d7-98263126abbe)

**Comment by AdamWr** [comment]
Shouldn't it be `||browserleaks.com^$permissions=geolocation=()` (missed `=` before `()`)?

**Comment by scripthunter7** [comment]
Checking if a regexp can be optimized. For example:
- `a{1,2}a{2,3}` -> `a{3,5}`
- `[0-9]` -> `\d`
- etc

Reference: https://github.com/DmitrySoshnikov/regexp-tree/tree/master/src/optimizer

**Comment by scripthunter7** [comment]
Deprecate old HTML filtering syntax, and suggest the modern approach. For example

- `tag-content` should be replaced by `contains`:
    ```adblock
    example.com$$script[tag-content="foo"]
    ```
    ↓↓↓ should be fixed as ↓↓↓
    ```adblock
    example.com$$script:contains(foo)
    ```
- `wildcard` should be replaced by `contains`, and glob pattern should be replaced by regular expression:
    ```adblock
    example.com$$script[wildcard="*banner*text*"]
    ```
    ↓↓↓ should be fixed as ↓↓↓
    ```adblock
    example.com$$script:contains(/banner.*text/)
    ```

**Comment by scripthunter7** [comment]
Rule for checking possible CSS selectors simplifications:
- `[id="foo"]` → `#foo`
- `[class="foo"]` → `.foo`
- `[id="foo"][class="bar"]` → `#foo.bar`
- `:not(:not(.foo))` → `.foo`

**Comment by piquark6046** [comment]
Unnecessary spaces 
https://github.com/AdguardTeam/AdguardFilters/commit/cf076573352cf5255d292b0c0d92e223b81b9216

**Comment by scripthunter7** [comment]
> Unnecessary spaces [AdguardTeam/AdguardFilters@cf07657](https://github.com/AdguardTeam/AdguardFilters/commit/cf076573352cf5255d292b0c0d92e223b81b9216)

@piquark6046 Its related to https://github.com/AdguardTeam/AGLint/issues/196

**Comment by piquark6046** [comment]
https://github.com/AdguardTeam/AGLint/issues/13#issuecomment-1914487255


I don't know that. :sweat_smile:

**Comment by Alex-302** [comment]
> * `[0-9]` → `\d`

`[0-9]` is used for iOS.

**Comment by Alex-302** [comment]
https://github.com/AdguardTeam/AGLint/issues/13#issuecomment-1913158450

- `[id="foo"]` → `#foo`
- `[class="foo"]` → `.foo`
- `[class="foo bar"]` → `.foo.bar`
- `[id="foo"][class="bar"]` → `#foo.bar`

This makes rules less specific. In some cases we using `[id="foo"]` to make rules less wide or dangerous.

**Comment by Yuki2718** [comment]
Maybe more than 2 `||` at the start of a rule can be detected? https://github.com/AdguardTeam/AdguardFilters/commit/59bbdaa0a0e044edfe7882d558aa8747c62513f8

**Comment by scripthunter7** [comment]
@Yuki2718 yes, its a part of validating network rule patterns, we plan to implement it there

**Comment by piquark6046** [comment]
Some spaces/a space between a `||` indicator and domain string.
E.g., `|| exmaple.com^`

**Comment by Yuki2718** [comment]
Maybe https://github.com/AdguardTeam/AdguardFilters/commit/cf9d8c9644446bfcac0ac93aecde55ab1ff8aeef#diff-9e4a61756ff9cff68c9ea921d901f0ff956d3f75b24dddbfc6ef1380c1a8242cR2130 could be detected? It looks AdGuard actually ignores the rule, but syntax-wise this should have disabled generic cosmetic filters on any site, which is the case for uBO.

**Comment by Alex-302** [comment]
Add pair checking for 
```
! aglint-disable [xxx]

! aglint-enable  [xxx]
```
